Zenoh - Load session configuration

caguero opened the pull request:
# 🎉 New feature

Closes #698

## Summary
<!--Explain changes made, the expected behavior, and provide any other additional
context (e.g., screenshots, gifs) if appropriate.-->

This patch adds the ability to load a Zenoh configuration file. The Zenoh [router](https://github.com/eclipse-zenoh/zenoh/tree/main/zenohd) is also supported.

Right now, you can set the configuration file with two methods:

1. Using the `ZENOH_CONFIG` environment variable. This is the option with the highest priority and we leverage [zenoh::Config::from_env()](https://zenoh-cpp.readthedocs.io/en/1.5.1/config.html#_CPPv4N5zenoh6Config8from_envEP7ZResult), which looks for this environment variable. That's why we don't need to create a custom environment variable for it.

2. If (1) isn't set, the default Zenoh configuration is used without any external file.

~~**Note**: The tests are expected to fail. See the Discussion section.~~

## Test it
<!--Explain how reviewers can test this new feature manually.-->

Follow the provided tutorial.

## Discussion

1. **Testing strategy.** ~~Right now, the default configuration files are based on ROS RMW Zenoh, which requires the Zenoh router to be running and multicast is disabled. Then, for testing, we should run the Zenoh router (not done in this patch) and distribute it. Alternatively, we could provide a separate configuration file with multicast enabled and not needing the Zenoh router. Another variant could be that our default configuration file has multicast enabled and doesn't require the router.~~ Using the default configuration allows us to run all tests without installing the Zenoh router or using special config files.

2. **Order in which we look for the configuration file.** ~~Right now, the default configuration file is the last resort. We first check for an environment variable, and then, for the existence of the user configuration file. If we don't find the user config file, we install it. Unless something goes wrong, it's not possible to use a default configuration, we'll almost always load a config file. We could consider other alternatives:~~ This is the alternative we're proposing:
  
   2.1 Default config first unless we explicitly set ZENOH_CONFIG.

~~2.2 Default config unless we explicitly set ZENOH_CONFIG or a config file is present at the default location. We could decide not to install the user config ourselves.~~

3. **Installable user config file.** ~~This patch installs a new user config file if it's not present in the default location (`~/.gz/transport`). My main concern is how to disitribute new versions of this file once the users already have an existing config file in their systems.~~ No instalable user config file.

~~I'm looking for some feedback on (1), (2) and (3).~~

~~I'm slightly leaning towards leaving this patch as it is but with the configuration file set with multicast enabled and not requiring the router. This way, the default experience is the same as right now, tests should work without extra work and users have an installed config file to play with and with the possibility of using the router if needed.~~

## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [x] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [ ] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers
- [ ] Was GenAI used to generate this PR? If so, make sure to add "Generated-by" to your commits. (See [this policy](https://osralliance.org/wp-content/uploads/2025/05/OSRF-Policy-on-the-Use-of-Generative-Tools-Generative-AI-in-Contributions.pdf) for more info.)

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` and `Generated-by` messages.

caguero commented on src/NodeSharedPrivate.hh:
```diff
@@ -110,7 +120,13 @@ namespace gz::transport
+    /// If the environment variable ZENOH_CONFIG is set, use that config file.
+    /// Otherwise, use the default Zenoh configuration.
+    /// \return The Zenoh configuration object.
+    public: GZ_TRANSPORT_VISIBLE zenoh::Config ZenohConfig();
```
Refactored in [f0c965c](https://github.com/gazebosim/gz-transport/pull/733/commits/f0c965c6eab47143380f67259464502d48d83093).

azeey commented on tutorials/24_zenoh_configuration.md:
```diff
@@ -0,0 +1,82 @@
+### Prerequisites
+
+First, install the Zenoh router by following the
+[official Zenoh installation guide](https://zenoh.io/docs/getting-started/installation/).
```
Is it okay to insteall from upstream when we have our verion in our apt repos? Does the potential version mismatch cause a problem?

azeey commented on tutorials/24_zenoh_configuration.md:
````diff
@@ -0,0 +1,82 @@
+```
+
+The subscriber should receive messages from the publisher. The router acts as
+the central point through which all messages flow.
````
Is this true? I thought the router in the default case would only help with discovery.

caguero commented on src/CMakeLists.txt:
```diff
@@ -38,7 +38,7 @@ endif()
 gz_build_tests(TYPE UNIT SOURCES ${gtest_sources}
   TEST_LIST test_list
-  LIB_DEPS ${EXTRA_TEST_LIB_DEPS} test_config)
+  LIB_DEPS ${EXTRA_TEST_LIB_DEPS} test_config ${ZeroMQ_TARGET})
```
We need it because the main library links privately with ZeroMQ, I don't think the transitive link is propagated to the test. And now, the new test includes a ZeroMQ header.

azeey commented on tutorials/24_zenoh_configuration.md:
```diff
@@ -0,0 +1,97 @@
+
+## Configuration Methods
+
+There are multiple ways to configure Zenoh when using Gazebo Transport:
```
nit; Multiple sounds like more than 2.
```suggestion
There are two ways to configure Zenoh when using Gazebo Transport:
```

caguero commented:
> It would be great to also add an example where data actually flows through the router.

Added in [9d60cac](https://github.com/gazebosim/gz-transport/pull/733/commits/9d60cacc09bc1ff1da5fe9421ffd904c09275d89). I simplified the config files to very minimum.

caguero commented on src/NodeSharedPrivate.hh:
```diff
@@ -83,8 +93,31 @@ namespace gz::transport
+          // wait for the router to become available.
+          std::cerr << "Failed to open Zenoh session: "
+                    << e.what() << std::endl;
+          std::exit(EXIT_FAILURE);
```
Replaced in [fe42288](https://github.com/gazebosim/gz-transport/pull/733/commits/fe42288291a3e5663e950cdde2aa4a4adccdb060).
